### PR TITLE
feat(command): add `tasks` argument

### LINKED ISSUE #43: Command to view all tasks
`npm run` lists all scripts available in `package.json`.  I often want to list all tasks available to `lab`.  Is there a way to do this?  If not, I'd propose something like `lab tasks` to view all tasks.

### PR DESCRIPTION
**Pull request by levithomason**
Fixes #43 

This PR adds a `lab tasks` command to view all tasks.  Things to note:

### No CLI commands

There is no current mechanism for CLI commands as all arguments are treated as tasks.  I've added a proposal for basic support for this.

### `tasks` vs `ls`
I've gone with `tasks`, however, if the sole purpose of DevLab will only be to run tasks then this could arguably be `lab ls` instead.

***

P.S

My thoughts on general architecture after working on DevLab.  This is food for thought for its future.

### Separation of concerns
In the larger picture, it would be nice to clean up the CLI bootstrap process.  I had to add some early returns in places as methods named things like `getConfig` actually do a lot more than get config (validation, execute commands, etc.)  It would be ideal to have `getConfig` strictly load the config or throw.  Similarly for all other methods.

### lib
Currently, the CLI and lib methods are baked together into objects.  It would be nice to separate all the methods into plain functions that can be called from anywhere and put them in a `/lib`.  Then, have a CLI that just wraps the `/lib`.

This would make testing the functions and lib much easier and also help with the separation of concerns.

### args
The `args` module does a lot more than process args, it also runs commands and includes multiple process.exit calls.  It would be preferred to have the args module parse args and nothing more.  Then, things that need args can get the args and do stuff with them.

### process.exit
It would be nice to have a [single process exit](https://www.nczonline.net/blog/2014/02/04/maintainable-node-js-javascript-avoid-process-exit/) in the entire CLI.  There would be one handler for when the CLI was done, it would exit with the right code.  This would be the top level promise chain allowing all underlying methods to throw if needed.

### output (i.e. logging)
There is an output module, however, it does not include a plain `log` method.  This means the cli uses the default console.log method.  This works, however, it cannot be turned off in tests or controlled in any other way.  It would be ideal to include a `log` method that could be turned on/off and also give standardized logging to DevLab.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,11 +26,10 @@
   },
   "preferGlobal": true,
   "scripts": {
-    "test": "npm run lint -s && npm run cover -s",
+    "test": "npm run lint -s && npm run test:cover -s",
+    "test:cover": "istanbul cover _mocha test/src/**/*.spec.js",
+    "test:watch": "nodemon --exec \"npm run test:cover || exit 1\"",
```
Comment (levithomason) [inline package.json]: Standardize to TAs conventions.

File: src/command.js [hunk 2]
```diff
@@ -27,7 +27,7 @@ const command = {
    */
   parseArgs: (type, args) => _.chain((item) => ([ command.args[type], command.parseHostEnvVars(item) ]), args),
   /**
-   * Parses config object and returns conatiner name. Will have dl_ prefix and
+   * Parses config object and returns container name. Will have dl_ prefix and
```
Comment (levithomason) [inline src/command.js]: Whrrps.

File: src/commands.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+const _ = require('redash')
+
+const pad = (len, str) => str.length < len ? str + ' '.repeat(len - str.length) : str
+
+const commands = {
+  /**
+   * List all available tasks
```
Comment (levithomason) [inline src/commands.js]: Do the thing I set out to do.

File: src/index.js [hunk 4]
```diff
@@ -70,18 +71,21 @@ const instance = {
    * @returns {object} promise
    */
   start: () => Promise.resolve().then(() => {
+    // CLI command, execute and return early
+    const cliCommand = commands[args.raw._[0]]
+    if (cliCommand) return cliCommand()
```
Comment (levithomason) [inline src/index.js]: Here is where I've shimmed in support for commands.  If there is a command available, it will be executed immediately and the usual task execution circumvented.  This is a bit naive but works for now.

File: test/src/commands.spec.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    })
+
+    after(() => {
+      console.log.restore()
+      fs.unlinkAsync(confPath)
+    })
+
+    it('pretty prints all tasks in the config', () => {
+      commands.tasks()
+      expect(console.log).to.have.been.calledWithExactly([
```
Comment (levithomason) [inline test/src/commands.spec.js]: Would be nicer here to spy on the `output` module's `log` method, if one existed.  I decided not to implement it in this PR as there are several other places using the console log.  If we decide to go with one, we can do a proper refactor in another PR.

**Comment by Fluidbyte** [comment]
If you'll notice other non-task commands (`--cleanup` and `--cleanup-all` are flags that get passed to the `lab` parent command there's good reason for that. Unlike most other CLI's where the available tasks are dictated for you, the whole point is the tasks and 99% of the usage of this tool is calling custom tasks, either singly (`lab test`) or chained (`lab clean install test`).

The architecture of the `args` module is built around this, the early returns and `process.exits` are specifically in place to prevent the CLI handling to seep into other areas. This as opposed to something like `npm` with a few reserved commands (`test`, `install`, `ls`...) then everything else (i.e. custom scripts) a child of `npm run <whatever>`.

This project has undergone several major overhauls and the architecture in place is specific to the fact that the one, primary goal, is simply to execute a task by building the massive `docker run ...` commands and, well, running them.

All this to say, I would prefer see the command handled by a flag like `lab --tasks` or `lab --ls` so it can't cause conflicts. This would ideally call a method in `utils`.

**Comment by levithomason** [comment]
Shall we go with `lab --tasks` or `lab --ls` then or keep the current `lab tasks`?  I had done this as an arg intially, however, it just felt like abuse to shove commands into args 😕

**Comment by Fluidbyte** [comment]
Yeah, I'm torn on this one. Your argument makes complete sense, I just feel like it leaves room for conflict, especially if it were `lab ls`. I think I'm alright with `lab tasks` as a reserved command as long as we move it into `args`.

**Comment by levithomason** [comment]
OK, I'll take a look at this tomorrow then 👍

**Comment by Fluidbyte** [comment]
After thinking about this I'm cool with either of these approaches:

1. Go simple; make this an argument `--tasks` and call it a day.

2. More complex; in `commands.js` create a `reservedTasks` list, add `tasks` and move the `--cleanup` and `--cleanup-all` here as well. Then execute `reservedTasks` the same as a custom task. Additionally, on run I would like to see this evaluate the `devlab.yml` and throw an error if someone tried to create a task with the same name as a reserved task.

I think option 1 is good because I don't see us expanding much on tasks beyond what the user creates. However, on the flip-side, if we do create reserved/built-in tasks we could expand this in the future. For example, I've been thinking it would be nice to have some built-in commands like `env` which would become very easy with option 2. We've also discussed the idea of devlab plugins, which could open the door to something like a `devlab-node` plugin that could apply reserved tasks like `npm install` (`install`) and `npm test` (`test`).

**Comment by levithomason** [comment]
I completely agree. Let's do 1 for now, since it is low hanging, and keep 2 as our longer term goal. Glad we found a meeting of minds here.  I'll push the update.

**Comment by levithomason** [comment]
Have a look at that and see if it is what you were thinking.

**Comment by Fluidbyte** [comment]
Clinton Approves
